testing/mock.ts

```diff
@@ -341,19 +358,24 @@ export interface Stub<
 /** Replaces an instance method with a Stub. */
 export function stub<
   Self,
-  // deno-lint-ignore no-explicit-any
-  Args extends unknown[] = any[],
-  Return = undefined,
->(self: Self, property: keyof Self): Stub<Self, Args, Return>;
+  Prop extends keyof Self,
+>(
+  self: Self,
+  property: Prop,
+): Stub<Self, GetParametersFromProp<Self, Prop>, GetReturnFromProp<Self, Prop>>;
 export function stub<
   Self,
-  Args extends unknown[],
-  Return,
+  Prop extends keyof Self,
 >(
   self: Self,
-  property: keyof Self,
-  func: (this: Self, ...args: Args) => Return,
-): Stub<Self, Args, Return>;
+  property: Prop,
+  func: (
+    this: Self,
+    // deno-lint-ignore no-explicit-any
+    ...args: any[]
+    // deno-lint-ignore no-explicit-any
+  ) => any,
```

Comment:
I think it would be good to keep the return type how you had it unless it caused issues. That would help prevent people from using stubs that return the wrong type of values.

Reply:
It's causing some issues in `time_test.ts` when trying to stub symbols on `Date.prototype`. But I see now it was getting type errors there anyway.
I'm not happy about the need to cast to `any` in order to fix it, but I'm not sure if there's a better way.

Reply:
I think that time test case just isn't needed. I made it to basically test that all the methods call through to the original, passing all the same arguments and returning same return value. I just used symbols since they were all different. I think it would be fine to just change that test to verifying one method and one symbol gets replaced correctly and we could change it to passing in the correct args instead of using symbols. Spy on the original date functions instead of stubbing them.

Reply:
Hmm not sure if I did this correctly, but I've added an extra commit that makes it only check if toISOString passes through to the original.

Reply:
Hey I just started using `stub()` in production tests and I think it could be pretty beneficial to have type checking for the stubbed function. Right now all parameters are basically `any` without causing an "explicit any" type error, so you don't get any warnings about it either.

Did you specifically want to allow for changing the types of arguments entirely, or were you mostly concerned about passing in a function with zero arguments? Because it seems like using the `GetParametersFromProp<Self, Prop>` from earlier (without `Partial`) still allows for that. I.e.:

```ts
const obj = {
    someFn(x: number, y: string) {}
}
stub(obj,"someFn", () => {})
```

works fine without any type errors.

The reason why I'm asking is because in JavaScript with JSDoc it's pretty tedious to add types to arrow functions that are immediately passed in other functions, since you have to declare the function separately and then pass it in.

So say you have `obj.someFn` with the signature `(x: number, y: string) => void`, and you want to stub it while maintaining argument types, in TypeScript this would look like.

```ts
stub(obj,"someFn", (x: number, y: string) => {
    // do something with x and y
})
```

In JavaScript with JSDoc this would look like:
```js
/**
 * @param {number} x
 * @param {string} y
 */
function replaceFn(x, y) {
    // do something with x and y
}
stub(obj,"someFn", replaceFn)
```

Reply:
I've removed `any[]` as arguments and I've added a few more tests to verify that the inferred types are correct. There's also a test to check if mocking a function without any arguments is possible.
But if you prefer to leave it as `any[]` please let me know.

Reply:
As long as the stub helpers like returnsNext will still work when stubbing an instance method that has args on all it's call signatures, your change is fine. It sounds like it would be fine since you said `stub(obj,"someFn", () => {})` didn't give you a type issue.

I can test your changes this weekend on a project of mine that uses spies and stubs and let you know if I find any issues.

Reply:
> As long as the stub helpers like returnsNext will still work

Ah I see, I'll add an extra test for that just in case.

> I can test your changes this weekend on a project of mine that uses spies and stubs

Sounds good 👍